x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java

```diff
@@ -853,75 +840,76 @@ private void putTransientIfNonExisting(String key, Object value) {
         }
     }
 
-    private ElasticsearchSecurityException denialException(
+    private ElasticsearchSecurityException runAsDenied(Authentication authentication, String action) {
```

Comment:
In principle, we could further refactor by pulling all the code below into a `ElasticsearchSecurityExceptionFactory` class that would look s.th. like this:

```java
class ElasticsearchSecurityExceptionFactory {
  private final boolean isAnonymousEnabled;
  private final boolean anonymousAuthzExceptionEnabled;
  private final AnonymousUser anonymousUser;
  private final AuthenticationFailureHandler authcFailureHandler;
  
  ElasticsearchSecurityException runAsDenied(Authentication authentication, String action) {
        ....
  }
  
  ElasticsearchSecurityException bulkItemActionDenied(
        Authentication authentication,
        String action,
        TransportRequest request,
        String resolvedIndex
    ) 

    ...
}
``` 

`denialException` and `AuthorizationDenialMessages` could be entirely encapsulated by that class.

However, that's a bigger undertaking - not sure if it's worth it.

Reply:
We can stop at extracting out the string composing part in this PR. I am OK with it.